File: 2023/day20/solution.py
#!/usr/bin/env python3
import sys
import re
from collections import deque
from typing import Self
from math import lcm  #actually it's called "maths", sweetie

sys.setrecursionlimit(1_000_000)
DEBUG = 0
FILENAME = 'input.txt'
PART1 = PART2 = True

# yapf: disable
for arg in sys.argv[1:]:
  if arg.startswith('-e'): FILENAME = f'''example{arg[2:] if len(arg)>2 else ''}.txt'''
  elif arg.startswith('-d'): DEBUG = int(arg[2:]) if len(arg) > 2 else 1
  else:
    match arg:
      case '-i': FILENAME = 'input.txt'
      case '-p1': PART1, PART2 = True, False
      case '-p2': PART1, PART2 = False, True
      case '-p0': PART1, PART2 = False, False
      case _: raise Exception(f'unknown {arg=}')

class State(): # ugh. but enums are worse
  LOW = False
  HIGH = True
  def __init__(self, val: bool|Self=False): self._val = bool(val)
  def flip(self): self._val ^= True; return self
  def set(self,val:bool): self._val=bool(val); return self
  def set_low(self): self._val=False; return self
  def set_high(self): self._val=True; return self
  def __bool__(self): return self._val
  def __str__(self): return 'HIGH' if self._val else 'LOW'
  def __repr__(self): return f'{self.__class__.__name__}.{self.__str__()}'
  def __eq__(self,other:Self|bool): return self._val==bool(other) #type: ignore
  def copy(self): return State(self._val)

class Module():
  FLIP_FLOP = '%'
  CONJUNCTION = '&'
  _types = {
    FLIP_FLOP: 'flip flop',
    CONJUNCTION: 'conjunction',
  }
  def __init__(self, name: str, type: str, targets: str):
    self._name = name
    self._type = type
    self._targets = tuple(targets.split(', '))
    self._state = State()
    self._inputs: dict[str, State] = {}
  def set_input(self, input: str, state: State):
    # only valid on conjunctions but i cba writing a base and subclasses
    self._inputs[input] = state
    if all(self._inputs.values()): self.state.set_low()
    else: self.state.set_high()
  @property # should be a frozendict really
  def inputs(self): return self._inputs
  @property
  def typename(self): return self._types.get(self._type, 'module')
  @property
  def name(self): return self._name
  @property
  def type(self): return self._type
  @property
  def targets(self): return self._targets
  @property
  def state(self): return self._state
  def __repr__(self): return f'<Module type={self.typename} targets={self.targets} inputs={self._inputs} state={self.state}>'
#yapf: enable

MODULE_RE = re.compile(r'^([%&])?([a-z]+) -> ([a-z, ]+)$', re.MULTILINE)
with open(FILENAME, 'r') as f:
  modules = {
    match.group(2): Module(
      match.group(2),
      match.group(1),
      match.group(3),
    )
    for match in MODULE_RE.finditer(f.read())
  }


def solve(iterations=1000, rx_trap=False):
  # reset initial states
  for module in modules.values():
    if module.type == Module.FLIP_FLOP: module.state.set_low()
    elif module.type == Module.CONJUNCTION:
      for input in modules.values():
        if module.name in input.targets:
          module.set_input(input.name, State())

  if DEBUG > 2:
    print(f'{modules=}')

  # rx has only one input
  rx_driver = [x.name for x in modules.values() if 'rx' in x.targets][0]
  # l1 didn't go low after 100k button presses
  rx_l1 = {x.name for x in modules.values() if rx_driver in x.targets}
  # so lets try l2. set because there might be many pulses per button press
  rx_traces: dict[str, set[int]] = {x.name: set() for x in modules.values() if any(y in x.targets for y in rx_l1)}
  rx_trace_state = State.LOW

  count_low, count_high = 0, 0
  plsbreak = False
  for i in range(iterations):
    inputs = deque([('button', 'broadcaster', State())])
    while inputs:
      sender_name, receiver_name, state = inputs.popleft()
      if state == State.HIGH: count_high += 1
      elif state == State.LOW: count_low += 1
      else: assert False, f'{state=}'
      receiver = modules.get(receiver_name)
      sender = modules.get(sender_name)
      if DEBUG > 1: print(f'{i=} {sender_name} -> {receiver_name} ({receiver.typename if receiver else None}): {state}')
      if rx_trap and sender_name in rx_traces.keys() and state == rx_trace_state:
        # this is a bit backwards since we're adding the trace when the pulse is received, not when it's sent. but all pulses must be resolved before next i so maybe it's fine
        rx_traces[sender_name].add(i)
        if DEBUG > 0:
          print('traces:')
          for k, v in rx_traces.items():
            print(f'  {k=} {len(v)=} {sorted(v)[:5]=}')
        if all(len(x) >= 5 for x in rx_traces.values()):
          plsbreak = True
          break
      if receiver is None: continue
      # match case treats % as a sql-like glob pattern
      if receiver.type == Module.FLIP_FLOP:
        if state == State.LOW:
          # flip state and pass new state to targets
          prev_state = receiver.state.copy()
          receiver.state.flip()
          if DEBUG > 2:
            print(f'  {receiver.typename} {prev_state} -> {receiver.state}')
          for x in receiver.targets:
            inputs.append((receiver_name, x, receiver.state.copy()))
        elif DEBUG > 2:
          print(f'  ignore {state}')
      elif receiver.type == Module.CONJUNCTION:
        # update inputs. class determines new state. new state to targets
        if DEBUG > 2:
          print('  prev inputs:')
          for k, v in receiver.inputs.items():
            print(f'    {k}: {v}')
          print(f'  update input {sender_name} to {state}')
        prev_state = receiver.state.copy()
        receiver.set_input(sender_name, state)
        if DEBUG > 2: print(f'  {receiver.typename}: {prev_state} -> {receiver.state}')
        for x in receiver.targets:
          inputs.append((receiver_name, x, receiver.state.copy()))
      else:
        # pass the state directly to targets
        for x in receiver.targets:
          inputs.append((receiver_name, x, state.copy()))
    if plsbreak: break

  if not rx_trap:
    product = count_low * count_high
    if DEBUG > 0: print(f'{count_low=} {count_high=} {product=}')
    return product

  deltas: set[int] = set()
  for k, v in rx_traces.items():
    sv = sorted(v)
    if DEBUG > 0: print(f'''{k=} {sv=}''')
    item_deltas = {y - x for x, y in zip(sv, sv[1:])}
    offset = sv[0]
    if DEBUG > 0: print(f'  {item_deltas=} {offset=}')
    assert len(item_deltas) == 1
    delta = item_deltas.pop()
    # thankfully the aoc people have been kind and the offsets are all delta-1. so we can just lcm. the -1 is because we start at ix 0 so can be ignored
    assert delta == offset + 1
    deltas.add(delta)
  cycle = lcm(*deltas)
  if DEBUG > 0: print(f'{deltas=} {cycle=}')
  return cycle


def part1():
  result = solve()
  print(f'part 1: {result}')
  # 883726240


def part2():
  # give it a silly number of iterations. we shouldn't hit that unless something's gone very wrong (foreshadowing)
  result = solve(100_000, True)
  print(f'part 2: {result}')
  # 211712400442661


if PART1 == PART2 == False: solve(5)
if PART1: part1()
if PART2: part2()

File: 2023/day20/solution.ts
#!/usr/bin/env -S deno --allow-read
// #region base aoc template
declare global {
  interface Math {
    gcd(left: number, right: number): number;
    lcm(values: number[]): number;
  }
}

Math.gcd = (left: number, right: number) => {
  while (right !== 0) [left, right] = [right, left % right];
  return left;
};
Math.lcm = (values: number[]) => values.reduce((acc, item) => (acc * item) / Math.gcd(acc, item), 1);

const Maths = Math;

class Args {
  constructor(public filename = 'input.txt', public debug = 0, public part1 = true, public part2 = true) {
    for (const arg of Deno.args) {
      const [key, value] = [arg.slice(0, 2), (fallback = 0) => Number(arg.slice(2) || fallback)];
      if (key === '-d') debug = value(1);
      else if (key === '-e') filename = `example${arg.slice(2)}.txt`;
      else if (key === '-i') filename = 'input.txt';
      else if (key === '-p') [part1, part2] = [(value(0) & 1) === 1, (value(0) & 2) === 2];
      else throw new Error(`unrecognised arg="${arg}"`);
    }
    [this.filename, this.debug, this.part1, this.part2] = [filename, debug, part1, part2];
    console.log(`args: {filename: "${filename}", debug: ${debug}, part1: ${part1}, part2: ${part2} }`);
  }
}

class Deque<T> {
  private ring: T[];
  private front: number = 0;
  private back: number = 0;
  constructor(public length: number) {
    this.ring = new Array<T>(length);
  }
  pushFront = (...items: T[]) => {
    for (const item of items) {
      this.front = (this.front - 1 + this.length) % this.length;
      if (this.front == this.back) throw new Error('bruh');
      this.ring[this.front] = item;
    }
  };
  pushBack = (...items: T[]) => {
    for (const item of items) {
      this.ring[this.back] = item;
      this.back = (this.back + 1) % this.length;
      if (this.front == this.back) throw new Error('bruh');
    }
  };
  popFront = () => {
    const item = this.ring[this.front];
    this.front = (this.front + 1) % this.length;
    return item;
  };
  popBack = () => {
    this.back = (this.back - 1 + this.length) % this.length;
    return this.ring[this.back];
  };
  get empty() {
    return this.front === this.back;
  }
}

// deno-lint-ignore no-explicit-any
const debug = (level: number, ...data: any[]) => {
  if (args.debug >= level) console.debug(...data);
};

const args = new Args();
const input = await Deno.readTextFile(args.filename);
// #endregion

type NodeType = '%' | '&' | undefined;
interface Node {
  type: NodeType;
  targets: Array<string>;
}
interface FlipFlop extends Node {
  state: boolean;
}
interface Conjunction extends Node {
  inputs: Map<string, boolean>;
}

const parseInput = () => {
  const regex = /^([%&])?([a-z]+) -> ([a-z, ]+)$/gm;
  const nodes = new Map<string, Node | FlipFlop | Conjunction>(
    input.matchAll(regex).map((match) => [
      match[2],
      {
        type: match[1] as NodeType,
        targets: match[3].split(', '),
      } as Node,
    ])
  );
  // populate flipflop state
  nodes
    .entries()
    .filter(([_, v]) => v.type === '%')
    .forEach(([k, v]) => nodes.set(k, { ...v, state: false } as FlipFlop));
  // populate conjunction inputs
  nodes
    .entries()
    .filter(([_, conjunction]) => conjunction.type === '&')
    .forEach(([conjunctionName, conjunction]) => {
      nodes.set(conjunctionName, {
        ...conjunction,
        inputs: new Map<string, boolean>(
          nodes
            .entries()
            .filter(([_, v]) => v.targets.includes(conjunctionName))
            .map(([k, _]) => [k, false])
        ),
      } as Conjunction);
    });
  debug(3, nodes);
  return nodes;
};

const simulate = (
  nodes: Map<string, Node | FlipFlop | Conjunction>,
  stateHook?: (sender: string, target: string, state: boolean) => void
) => {
  interface Event {
    sender: string;
    target: string;
    state: boolean;
  }
  //FIXME: update Deque to grow dynamically. maybe start at 1000 and double in size whenever it becomes full?
  const queue = new Deque<Event>(100);
  queue.pushBack({ sender: 'button', target: 'broadcaster', state: false });
  while (!queue.empty) {
    const event = queue.popFront();
    if (typeof stateHook !== 'undefined') stateHook(event.sender, event.target, event.state);
    const node = nodes.get(event.target);
    if (typeof node === 'undefined') continue;
    if (node.type === '%') {
      // flip flop - low flips, high ignored
      if (event.state === false) {
        (node as FlipFlop).state = !(node as FlipFlop).state;
        queue.pushBack(...node.targets.map((target) => ({ sender: event.target, target, state: (node as FlipFlop).state })));
      }
    } else if (node.type === '&') {
      // conjunction. all high -> send low. otherwise, send high
      (node as Conjunction).inputs.set(event.sender, event.state);
      const sendState = !(node as Conjunction).inputs.values().reduce((acc, item) => acc && item, true);
      queue.pushBack(...node.targets.map((target) => ({ sender: event.target, target, state: sendState })));
    } else {
      queue.pushBack(...node.targets.map((target) => ({ sender: event.target, target, state: event.state })));
    }
  }
};

const part1 = () => {
  const nodes = parseInput();
  let [lowCount, highCount] = [0, 0];
  for (let i = 0; i < 1000; ++i)
    simulate(nodes, (_s, _t, state) => {
      if (state) ++highCount;
      else ++lowCount;
    });
  debug(1, { lowCount, highCount });
  console.log('part 1:', lowCount * highCount);
};

const part2 = () => {
  const nodes = parseInput();
  const rxL1 = new Set(
    nodes
      .entries()
      .filter(([_, v]) => v.targets.includes('rx'))
      .map(([k, _]) => k)
  );
  debug(1, { rxL1 });
  // watch rx's l2 drivers
  const watchNodes = new Map<string, Array<number>>(
    nodes
      .entries()
      .filter(([_, v]) => rxL1.intersection(new Set(v.targets)).size > 0)
      .map(([k, _]) => [k, new Array<number>()])
  );
  debug(1, { watchNodes });
  let i = 0;
  // loop until we have 5 values for all watched nodes
  while (watchNodes.values().some((item) => item.length < 5)) {
    i++;
    simulate(nodes, (sender, _, state) => {
      if (state && watchNodes.has(sender)) watchNodes.get(sender)!.push(i);
    });
  }
  debug(1, watchNodes);
  // map cycle number to deltas
  const deltas = watchNodes
    .values()
    .map((history) => history.map((item, i, arr) => arr[i + 1] - item).filter((_, i, arr) => i < arr.length - 1))
    .toArray();
  debug(1, { deltas });
  // assert that each node has only one distinct delta value
  if (deltas.some((item) => new Set<number>(item.values()).size > 1)) throw new Error('bruh');
  const firstDeltas = deltas.map((item) => item[0]);
  debug(1, { firstDeltas });
  // then lcm the deltas
  const cycleCount = Maths.lcm(firstDeltas);
  debug(1, { cycleCount });
  console.log('part 2:', cycleCount);
};

if (args.part1) part1();
if (args.part2) part2();
